dpnp/backend/extensions/indexing/choose.cpp

```diff
@@ -0,0 +1,463 @@
+            chc_ptrs, chc_offsets, n_chcs);
+    }
+    else {
+        auto src_strides = src.get_strides_vector();
+        auto dst_strides = dst.get_strides_vector();
+
+        pack_deps = _populate_choose_kernel_params(
+            exec_q, host_task_events, packed_chc_ptrs.get(),
+            packed_shapes_strides.get(), packed_chc_offsets.get(), src_shape,
+            sh_nelems, src_strides, dst_strides, chc_strides, chc_ptrs,
+            chc_offsets, n_chcs);
+    }
+
+    std::vector<sycl::event> all_deps;
+    all_deps.reserve(depends.size() + pack_deps.size());
+    all_deps.insert(std::end(all_deps), std::begin(pack_deps),
+                    std::end(pack_deps));
+    all_deps.insert(std::end(all_deps), std::begin(depends), std::end(depends));
+
+    sycl::event choose_generic_ev =
+        fn(exec_q, nelems, n_chcs, sh_nelems, packed_shapes_strides.get(),
+           src_data, dst_data, packed_chc_ptrs.get(), src_offset, dst_offset,
+           packed_chc_offsets.get(), all_deps);
+
+    // release usm_unique_ptrs
+    auto chc_ptrs_ = packed_chc_ptrs.release();
+    auto shapes_strides_ = packed_shapes_strides.release();
+    auto chc_offsets_ = packed_chc_offsets.release();
+
+    // free packed temporaries
+    sycl::event temporaries_cleanup_ev = exec_q.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(choose_generic_ev);
+        const auto &ctx = exec_q.get_context();
+
+        using dpctl::tensor::alloc_utils::sycl_free_noexcept;
+        cgh.host_task([chc_ptrs_, shapes_strides_, chc_offsets_, ctx]() {
+            sycl_free_noexcept(chc_ptrs_, ctx);
+            sycl_free_noexcept(shapes_strides_, ctx);
+            sycl_free_noexcept(chc_offsets_, ctx);
+        });
+    });
+
+    host_task_events.push_back(temporaries_cleanup_ev);
+
+    using dpctl::utils::keep_args_alive;
+    sycl::event arg_cleanup_ev =
+        keep_args_alive(exec_q, {src, py_chcs, dst}, host_task_events);
+
+    return std::make_pair(arg_cleanup_ev, choose_generic_ev);
+}
+
+template <typename fnT, typename IndT, typename T>
+struct ChooseWrapFactory
+{
+    fnT get()
+    {
+        if constexpr (std::is_integral<IndT>::value &&
+                      !std::is_same<IndT, bool>::value) {
+            using dpctl::tensor::indexing_utils::WrapIndex;
+            fnT fn = kernels::choose_impl<WrapIndex<IndT>, IndT, T>;
+            return fn;
+        }
+        else {
+            fnT fn = nullptr;
+            return fn;
+        }
+    }
+};
+
+template <typename fnT, typename IndT, typename T>
+struct ChooseClipFactory
+{
+    fnT get()
+    {
+        if constexpr (std::is_integral<IndT>::value &&
+                      !std::is_same<IndT, bool>::value) {
+            using dpctl::tensor::indexing_utils::ClipIndex;
+            fnT fn = kernels::choose_impl<ClipIndex<IndT>, IndT, T>;
+            return fn;
+        }
+        else {
+            fnT fn = nullptr;
+            return fn;
+        }
+    }
+};
```

Comment:
we can reduce the code duplication, like:
```cpp
template <typename fnT, typename IndT, typename T, typename Index>
struct ChooseFactory
{
    fnT get()
    {
        if constexpr (std::is_integral<IndT>::value &&
                      !std::is_same<IndT, bool>::value) {
            fnT fn = kernels::choose_impl<Index, IndT, T>;
            return fn;
        }
        else {
            fnT fn = nullptr;
            return fn;
        }
    }
};

using dpctl::tensor::indexing_utils::WrapIndex;
using dpctl::tensor::indexing_utils::ClipIndex;

template <typename fnT, typename IndT, typename T>
using ChooseWrapFactory = ChooseFactory<fnT, IndT, T, WrapIndex<IndT>>;

template <typename fnT, typename IndT, typename T>
using ChooseClipFactory = ChooseFactory<fnT, IndT, T, ClipIndex<IndT>>;
```

Reply:
Good idea